### ISSUE TITLE
Removal of `dealerdirect/phpcodesniffer-composer-installer` in a new major?

### ISSUE THREAD

**Issue by Ocramius** [slevomat/coding-standard#1413]
Let's face it: composer plugins are a pain.

I'd better not have them at all, and run everything with `--no-plugins` during installation.

I wonder if it's possible to remove `dealerdirect/phpcodesniffer-composer-installer` in the next major release, and to endorse (in the docs) referencing the sniff via absolute `vendor/` path?

**Comment by roslov** [comment]
You can use it like below:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="project-code-styling">
    <description>Local Project Coding Standard</description>

    <!-- Excludes vendor and temporary folders -->
    <exclude-pattern>vendor/*</exclude-pattern>
    <exclude-pattern>var/*</exclude-pattern>
    <exclude-pattern>tests/_support/_generated/*</exclude-pattern>

    <!-- Base rules -->
    <rule ref="vendor/slevomat/coding-standard/SlevomatCodingStandard" />

    <!-- Rules for tests -->
    <rule ref="PSR2.Methods.MethodDeclaration.Underscore">
        <exclude-pattern>tests/*</exclude-pattern>
    </rule>

    <!-- Other excluded rules -->
    <rule ref="SlevomatCodingStandard.TypeHints.DeclareStrictTypes.DeclareStrictTypesMissing">
        <exclude-pattern>config/bundles.php</exclude-pattern>
    </rule>
</ruleset>
```

I do not use that plugin too.

**Comment by Ocramius** [comment]
Indeed, what I'm wondering is if that should be the advertised installation method, and I'd like to get -1 dependencies in all my stack (by removing the plugin)

**Comment by kukulich** [comment]
It was added here: https://github.com/slevomat/coding-standard/pull/698

**Comment by Ocramius** [comment]
@jrfnl can this trend be reverted? WDYT?

The entire post-install-plugins stuff is such a massive chore, especially with `composer/composer:2.3.8` onwards (I can't anymore @_@ )

**Comment by jrfnl** [comment]
> Let's face it: composer plugins are a pain.

@Ocramius All I see here is a pet peeve, not a technical argument.

I would recommend to *not* remove the plugin.

Yes, people can manually register standards with PHPCS, but most people find that a pain, which is exactly what the plugin solves.

Yes, you can include the standard via a file reference, however, as the Slevomat standard is a "pick and choose" standard, this means you will always need to exclude a sh*tload of sniffs if you do that, which makes working with a custom ruleset really fiddly and also makes upgrading fiddly as you will normally not want to automatically include any and all newly added sniffs, so (nearly) every update of the standard would mean that the ruleset will need to be updated with additional excludes.

Yes, if you don't include the standard, you can include individual sniffs via file references, but you won't be able to use the `Stnd.Category.Sniff` references (as used in the README) anymore and will need to make the translation from sniff name to file reference yourself (and this may not always be as portable across OSes/platforms).

Without the standard being included (either via a file reference or via the standard name) you also won't be able to include/exclude complete categories of sniffs anymore.

All in all, it makes working with rulesets a lot more fiddly if the standard is not registered and that registration is the thing the plugin does...

> The entire post-install-plugins stuff is such a massive chore, especially with composer/composer:2.3.8 onwards

Curious what you mean. I don't recall seeing any problems in any of the projects I'm involved with.

Might be better to improve the installation instructions for the standard, like I've done in all standards I maintain. Those instructions now look something like this:
```bash
composer config allow-plugins.dealerdirect/phpcodesniffer-composer-installer true
composer require --dev phpcsstandards/phpcsextra:"^1.0"
```

**Comment by Ocramius** [comment]
> Curious what you mean.

Currently restoring CI for ~100+ repositories under the @laminas organisation, as well as mine and @roave's. @psr7-sessions broke too.

In practice, the technical argument was to reduce dependencies, leading therefore to less coupling with the (high) volatility of composer's API.

I was unaware of the fact that referencing the XML means "all or nothing": how would a consumer of this package pick individual sniffs, if they had to do it without the plugin? 🤔

**Comment by Ocramius** [comment]
BTW, the breakage is in composer, which becomes a cyclic dependency between CI and PHPCS.

My hope is to remove reliance on anything that isn't necessary: from my end, I would totally be fine with maintaining a larger XML myself, in exchange for better stability.

Less dependencies is something to strive for, heh.

**Comment by roslov** [comment]
For my projects, I create a custom standard that I use:

https://packagist.org/packages/roslov/psr12ext

It contains all needed rules for me, mostly based on PSR-12 and Slevomat.

To use it, I have to create a pretty small `ruleset.xml`. Also, you can install in both ways — either via direct link to the standard or using the plugin.

So you can use a similar approach in your project. Or use my standard :-)

**Comment by Ocramius** [comment]
Yeah, I rely on this package for the sniffs, not for the ruleset: the plugin gets dragged in as a (very problematic) side-effect

**Comment by jrfnl** [comment]
> Currently restoring CI for ~100+ repositories under the https://github.com/laminas organisation, as well as mine and https://github.com/Roave's. https://github.com/psr7-sessions broke too.

I feel your pain. What broke ? Was it the `allow-plugins` requirement or something else ?

> less coupling with the (high) volatility of composer's API.

Understood. While it doesn't "solve" this, I can ensure you that the PHPCS Composer plugin is diligently being tested as soon as new Composer RCs are being released, so the chances of breakage due to the plugin _itself_ should be small.

Side-note: while I added the plugin in #698, I added it at the time as a `dev` requirement for internal/CI use for this standard.
It was moved to `require` by @kukulich in 9e6958cf59b726062a4046c5e6abf342dd3a00ef .

> how would a consumer of this package pick individual sniffs, if they had to do it without the plugin?

Okay, guess a little explanation of PHPCS basics is needed here.

The "base" functionality presumes all (external) standards will be registered with PHPCS using the `--runtime-set installed_paths ...` command.
This can be fiddly as any previous set paths will be overwritten whenever you run that command, so you have to be careful to make sure all paths to external standards are still registered.
👉🏻 This is what the plugin does - it traverses the Composer `vendor` directory of a project and registers the external PHPCS standards it finds with PHPCS all in one go (and automatically updates those paths whenever needed due to changed/updated dependencies).

When a standard is registered with PHPCS, you can refer to it by name.
```xml
<ruleset name="Example">
    <!-- Include a complete standard, i.e. this will include all sniffs from the standard + anything else referenced in the `ruleset.xml` file. -->
    <rule ref="SlevomatCodingStandard"/>

    <!-- Include all sniffs from a specific category in a standard. -->
    <rule ref="SlevomatCodingStandard.Classes"/>

    <!-- Include an individual sniff from a standard. -->
    <rule ref="SlevomatCodingStandard.Classes.EmptyLinesAroundClassBraces"/>
</ruleset>
```

When a standard is **not** registered with PHPCS, you can include it via a ruleset file reference.
```xml
<ruleset name="Example">
    <!-- Include a complete standard, i.e. this will include all sniffs from the standard + anything else referenced in the `ruleset.xml` file. -->
    <rule ref="vendor/slevomat/coding-standard/SlevomatCodingStandard/ruleset.xml" />

    <!-- Include all sniffs in a directory. -->
    <rule ref="vendor/slevomat/coding-standard/SlevomatCodingStandard/" />
</ruleset>
```

When a standard is **not** registered with PHPCS, you can also include individual sniffs by file reference.
```xml
<ruleset name="Example">
    <!-- Include an individual sniff from a standard. -->
    <rule ref="vendor/slevomat/coding-standard/SlevomatCodingStandard/Classes/EmptyLinesAroundClassBracesSniff" />
</ruleset>
```

However, something like the below will not work:
```xml
<ruleset name="Example">
    <rule ref="./vendor/slevomat/coding-standard/SlevomatCodingStandard/Sniffs/Classes/"/>

    <rule ref="SlevomatCodingStandard.Classes.EmptyLinesAroundClassBraces">
        <properties>
            <property name="linesCountAfterOpeningBrace" value="0"/>
            <property name="linesCountBeforeClosingBrace" value="0"/>
        </properties>
    </rule>
</ruleset>
```
... and will result in:
```
ERROR: Referenced sniff "SlevomatCodingStandard.Classes.EmptyLinesAroundClassBraces" does not exist
```

Does that help ?

Generally speaking, as rulesets are normally committed to a repo and used by devs locally on various OSes and also in CI, registering a standard and referencing sniffs using the `Stnd.Category.Sniff`-like syntax will be completely stable, while using file paths may not be as portable in all cases and breaks more easily.

For more details, also see: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-Ruleset

**Comment by kukulich** [comment]
> Side-note: while I added the plugin in #698, I added it at the time as a dev requirement for internal/CI use for this standard.
It was moved to require by @kukulich in 9e6958c .

Yes, see discussion https://github.com/slevomat/coding-standard/commit/9e6958cf59b726062a4046c5e6abf342dd3a00ef

**Comment by Ocramius** [comment]
## Recap on why plugins are trouble

 > Was it the `allow-plugins` requirement or something else ?

Currently, yes, but that's only the first symptom at this date.
 
There's coupling between what is in `composer.lock` and the currently running `composer/composer` version, which isn't part of the locked versions.

The problem with plugins at large, especially in the close future, is that upstream `composer/composer` will likely no longer introduce new major releases (for political reasons), but rather introduce "acceptable" BC breaks (makes me furious, but it is what it is, and they got enough trouble to deal with) in minor versions, after looking around for what plugins are doing, and "helping them transition".

The locked dependencies (including plugins) are still there, and `--prefer-lowest` installations still need to be tested, for the sake of downstream consumers. This means that installing a newer `composer/composer` (due to how most CI pipelines work) will break fairly often, as it will become incompatible with stable (semver-compliant) plugin versions too.

This means that every time composer releases something, we have hundreds of broken CI runs, especially problematic for LTS (and feature-complete) packages that shouldn't really be touched.

The alternative is to stick to composer LTS, which means that plugins don't get tested against newer composer versions either (more problems).

My approach is to cut the problem at its root, removing the problematic bidirectional dependency, rather than finding more workarounds. 

## The problem with running/configuring a plugin-less PHP_CodeSniffer setup

 > However, something like the below will not work:
 >
 > `<snip> section explaining conflict between importing standard by path and configuring rules for it </snip>`

Ok, this brings in some clarity, thanks @jrfnl.

I'm mostly wondering if this is solvable (architecturally) within PHP_CodeSniffer by having a convention on how standards could be loaded with an existing autoloader, and without `--runtime-set installed_paths ...`?

Relying on autoloader semantics would remove path lookups, allow the `Foo.Bar.Sniff` syntax (`s/./\\`, basically), and get rid of this issue, although I am not deep enough in the PHP_CodeSniffer design to bring valuable suggestions other than wishful thinking, and it may be out of scope for this specific issue. 

This can be done by hijacking the autoloader, which is something I have experience with, but let me know if you'd rather discuss it elsewhere (or if there's somebody that already approached this problem somewhere).

**Comment by jrfnl** [comment]
### Re: Composer vs this plugin

This specific plugin supports Composer 1.x all the way up to current and handles BC-breaks in Composer internally in the plugin.
This is safeguarded via the CI for the plugin and the intention is to continue to support this.
See: https://github.com/PHPCSStandards/composer-installer/blob/master/.github/workflows/integrationtest.yml#L23-L100 and https://github.com/PHPCSStandards/composer-installer/actions/runs/2714142372

While that may not be the case for other plugins and I understand your gripe with Composer for that, this plugin is not expected to have a problem with different Composer versions.

Even so, I'd encourage and support having a discussion in the Composer repo about semver and breaking changes.... 

### Re: PHPCS autoloading vs external standards

I think that should be discussed in the PHP_CodeSniffer repo. Based on my experience contributing there, any changes would probably only be accepted for PHPCS 4.x (ETA unknown, may still be a few years).

PHPCS currently uses a custom autoloader, which also handles the autoloading for external standards and does the translation between class name and sniff code.
While PHPCS also supports for external standards to register an additional autoloader and will also look to a potentially available Composer `vendor/autoload.php` file, the translation part will generally be broken for sniffs not loaded via the custom autoloader, which limits the supported functionality in PHPCS.

Some additional context: PHPCS is an "old" project and still supports PEAR installs as well. PEAR support will be dropped in PHPCS 4.x, which may open the way to change this, but I'd need to dig in deep to check this through.

All the same, any change which would remove/reduce support for the sniff codes with dots would be a HUGE BC-break and will be unlikely to be accepted.

**Comment by jrfnl** [comment]
P.S.: also note that making any changes to this functionality in PHPCS is **_hard_** due to the lack of test for the Core classes. I've slowly been adding tests for anything I touch in the Core classes, but there are still huge chunks of Core functionality which are not covered by tests.

**Comment by Ocramius** [comment]
> This specific plugin supports Composer 1.x all the way up to current and handles BC-breaks in Composer internally in the plugin.

Yeh, it's mostly about locked ranges, and the constant little fractures between dependencies not declared in these lockfiles (composer, in this case).


 > Even so, I'd encourage and support having a discussion in the Composer repo about semver and breaking changes....

I already had a direct discussion some time ago, and it's a project direction/decision that was already taken, and will ultimately make me drop composer as an installer in the future (also here, few years in the future, since it takes time and work).


 > I think that should be discussed in the PHP_CodeSniffer repo. Based on my experience contributing there, any changes would probably only be accepted for PHPCS 4.x (ETA unknown, may still be a few years).

Works for me. I will continue my busy-work in playing catch-up with dependencies, and think the problem through.

 > Some additional context: PHPCS is an "old" project and still supports PEAR installs as well. PEAR support will be dropped in PHPCS 4.x, which may open the way to change this, but I'd need to dig in deep to check this through.

Well aware of it 👍 

 > All the same, any change which would remove/reduce support for the sniff codes with dots would be a HUGE BC-break and will be unlikely to be accepted.

I think this is where a hacky autoloader could help: this is something I've already implemented in https://github.com/Roave/BetterReflection/blob/40fd8e5bfb279051e28fecbb9adb9513039f039a/src/SourceLocator/Type/AutoloadSourceLocator.php#L130-L198 before. In practice, as long as somebody (anybody) has configured an autoloader, we can locate a class-alike symbol in the sources, with some performance overhead, obviously :-) 

If you think this is a possible/viable solution, and you aren't horrified too much by what you just looked at, I could propose it for upstream. 

 > P.S.: also note that making any changes to this functionality in PHPCS is hard due to the lack of tests for the Core classes. 

Different times, different standards: fully relatable.

**Comment by jrfnl** [comment]
> If you think this is a possible/viable solution, and you aren't horrified too much by what you just looked at, I could propose it for upstream.

@Ocramius This really isn't up to me (I haven't got commit) and I haven't seen a roadmap published of what Greg intends to include in PHPCS 4.x, so I can't advise you on whether or not this would be something which would be considered for inclusion.
I'm definitely not horrified by that code though. Makes a lot of sense to me (in that context).

However, if I look at PHPCS 4.x, especially with PEAR support now being dropped, I would probably strongly advocate for adding native support for Composer installed external standards to PHPCS. That would make the plugin redundant.

Notes:
* It would still also need to support manually set `installed_paths` and not overwrite these as AFAIK stand-alone install/PHAR based use (Phive) with external standards should still remain supported.
* Greg is generally against adding (non-dev) external dependencies to PHPCS, so any solution would need to be PHPCS native code-wise.

Alternatively, adding support for saving the `CodeSniffer.conf` file (which is where the `installed_paths` setting is saved) in a project root (and which would allow for putting it under version control) would also be an option.

Currently, you can already set the `installed_paths` from within a custom ruleset. This overrules anything set for `installed_paths` in the `CodeSniffer.conf` file.

Both these last two options, however, are still subject to the following two problems:
1. In my experience this is not very stable/portable across OSes with relative vs absolute paths etc.
2. If an external standard starts to ship with a dependency on another external standard, changes the project directory layout (etc etc), you manually have to adjust the setting.

Some loosely related issues upstream:
* https://github.com/squizlabs/PHP_CodeSniffer/issues/2062
* https://github.com/squizlabs/PHP_CodeSniffer/issues/1436
* https://github.com/squizlabs/PHP_CodeSniffer/issues/1476

**Comment by Ocramius** [comment]
> That would make the plugin redundant.

That's the hope

 > Greg is generally against adding (non-dev) external dependencies to PHPCS, so any solution would need to be PHPCS native code-wise.

That would be the case, yes :-)

> 1. In my experience this is not very stable/portable across OSes with relative vs absolute paths etc.
> 2. If an external standard starts to ship with a dependency on another external standard, changes the project directory layout (etc etc), you manually have to adjust the setting.

Luckily, the approach above is mostly filesystem-agnostic 👍 

Thanks for the references: going to close here, since this entire issue would require rework in PHP_CodeSniffer, rather than here anyway. Thanks for the extremely useful insights: I have more material to sleep over and get ideas from 💪

**Comment by jrfnl** [comment]
@Ocramius If you think it would help, I'd be happy to work on this together at some point. Feel free to DM me on Twitter.

**Comment by Potherca** [comment]
As one of the original authors of `dealerdirect/phpcodesniffer-composer-installer`, regarding:

> That would make the plugin redundant.

When we originally created the plugin, the whole idea was for the project to (eventually) become redundant. It was an attempt to fix the "automatically install code sniffs" once and for all, as all solutions at the time had one or more severe flaws.

We originally tried going the [`composer/installers`](https://github.com/composer/installers/issues/332) route, but the fact that it took 2 years to get a simple yes/no question answered did not make us very hopeful of that route.

There was also [a discussion on adding the plugin in the Codesniffer project](https://github.com/squizlabs/PHP_CodeSniffer/pull/936). The conclusion at the time (2016 - 2019) was:

> The general issue of installing standards via composer may be addressed in the future, but not in version 3, and probably only after a full refactor and rethink about how standards work.

But things might have changed regarding the v4 release.

**Comment by marekdedic** [comment]
Hi,
I have just encountered this issue when my CI "randomly" broke (in fact, it was because GH actions upgraded composer and the new composer requires explicit plugin approval).

Looking at other phpcs standards such as `phpcompatibility` or `wpcs`, it seems that they instead opt to use the composer `suggests` field - they suggest using it because it makes it easier to use the standard. However, as the plugin is not required, it IMHO should not be a strict dependency, rather a suggestion...

**Comment by jrfnl** [comment]
@marekdedic Both PHPCompatibility as well as WPCS will actually require the plugin as of the next major release (work on both is ongoing) and support for non-Composer installs will no longer be provided.
Once these majors are released the install instructions for both will be updated to include the `composer config allow-plugins.dealerdirect/phpcodesniffer-composer-installer true` command, just like [PHPCSExtra](https://github.com/PHPCSStandards/PHPCSExtra#composer-project-based-installation) already does.

**Comment by marekdedic** [comment]
Oh, could you please maybe provide a bit more context as to why? (link to any issue etc.) Because it seems to me to only complicate using them, but I don't know the background enough to properly judge that :/

**Comment by jrfnl** [comment]
@marekdedic The background is quite simple: both will require one or more additional external standards as of the next release - PHPCompatibility => PHPCSUtils, WPCS => PHPCSUtils + PHPCSExtra -.

This means that the `installed_paths` for all installs will need to be updated and updated correctly and I have seen too many support requests about `installed_paths` already.

In other words: in both cases, the maintainers (not just me, this was discussed with all involved maintainers) have made a choice to rather spend their time on improving sniffs/new sniff than to spends (significant) time on answering support questions about `installed_paths`.

**Comment by jrfnl** [comment]
Note: non-Composer based installs will still be possible, it's just that we won't be providing support for it anymore.

**Comment by marekdedic** [comment]
Thanks for the explanation, now it makes sense :)
